### PR TITLE
[3.3.5] Social Links rewrite

### PR DESCRIPTION
**Pull request by Treeston**
Because I don't have enough PRs open yet.

Current SocialMgr immediately saves everything to database on change. That's bad.
It currently also iterates over all online players whenever somebody logs in or out. That's also pretty bad.

I toss the whole thing and rewrite it, and yes this is starting to be a bit of a trend by now.

SocialMgr now stores all social links authoritatively, which means it can also consistently maintain reverse lookups despite DB data only being saved periodically. Each player has pointers to its own maps so we don't even need to do hashmap lookups most of the time.

Enjoy.

### CODE REVIEW

**Comment by ccrs** [comment]
> I toss the whole thing and rewrite it, and yes this is starting to be a bit of a trend by now.

👍

File: src/server/game/Entities/Player/SocialMgr.cpp [hunk 1]
```diff
@@ -17,319 +17,332 @@
...
+    ASSERT(isFriend || (flag == SOCIAL_FLAG_IGNORED), "Invalid social flag %u to AddSocialLink", uint32(flag));
 
-    if (!itr->second.Flags)
+    WorldSession* playerSession = player->GetSession();
+    FriendsResult result;
+    Player const* targetPlayer = nullptr;
+    CharacterCacheEntry const* target = sCharacterCache->GetCharacterCacheByName(targetName);
     {
-        PreparedStatement* stmt = CharacterDatabase.GetPreparedStatement(CHAR_DEL_CHARACTER_SOCIAL);
+#define SEND_RESULT(v1,v2) do { result = (isFriend ? v1 : v2); goto send; } while(0);
```
Comment (ccrs) [inline src/server/game/Entities/Player/SocialMgr.cpp]: ummm what about defining a std::function localy? that goto should burn 🔥

File: src/server/game/Entities/Player/Player.h [hunk 2]
```diff
@@ -2386,7 +2389,9 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
         uint32 m_questRewardTalentCount;
 
         // Social
-        PlayerSocial* m_social;
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.h]: This is a bad change, don't bloat `Player` class

Comment (lachtanek) [reply src/server/game/Entities/Player/Player.h]: Maybe you could've used the old PlayerSocial, just improve/refactor/optimize it, seems better to me than throwing new methods/properties onto this huge thing that calls itself Player

**Comment by jackpoz** [comment]
https://github.com/TrinityCore/TrinityCore/issues/23825 might be related to this PR

File: src/server/game/Entities/Player/Player.h (hunk 2 above)
Comment (jackpoz) [reply src/server/game/Entities/Player/Player.h:2389]: Marking this PR as Feedback-FixOutdatedMissingWIP until this change is reverted

**Comment by Shauren** [comment]
Closing as abandoned and known to cause crashes, #23825

**Pull request closed without merging**